File: archive/Backend/app/routes/vehicles.py
from __future__ import annotations

import logging
from datetime import datetime, timezone

from bson import ObjectId
from bson.errors import InvalidId
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.config import settings
from app.services.mongodb_service import get_database

router = APIRouter(tags=["Vehicles"])
logger = logging.getLogger(__name__)

VEHICLE_TYPES = {"SUV", "MUV", "Van", "Tempo Traveller", "Bus", "Sedan"}


def _serialise(doc: dict) -> dict:
    return {
        "id":               str(doc["_id"]),
        "make":             doc.get("make", ""),
        "name":             doc.get("vehicle_name", ""),
        "vehicleNumber":    doc.get("number_plate", ""),
        "type":             doc.get("vehicle_type", "MUV"),
        "capacity":         doc.get("capacity", 7),
        "assignedDriverId": doc.get("assigned_driver_id"),
    }


def _oid(id_str: str) -> ObjectId:
    try:
        return ObjectId(id_str)
    except InvalidId:
        raise HTTPException(status_code=400, detail="Invalid vehicle id")


class VehicleCreate(BaseModel):
    make: str
    name: str
    vehicleNumber: str
    type: str = "MUV"
    capacity: int = 7


class VehicleUpdate(BaseModel):
    make: str | None = None
    name: str | None = None
    vehicleNumber: str | None = None
    type: str | None = None
    capacity: int | None = None
    assignedDriverId: str | None = None


@router.get("/")
async def list_vehicles():
    db = get_database()
    docs = await db[settings.vehicles_collection].find({}).to_list(None)
    return [_serialise(d) for d in docs]


@router.post("/", status_code=201)
async def create_vehicle(body: VehicleCreate):
    if body.type not in VEHICLE_TYPES:
        raise HTTPException(status_code=422, detail=f"type must be one of {sorted(VEHICLE_TYPES)}")
    now = datetime.now(timezone.utc)
    doc = {
        "make":               body.make.strip(),
        "vehicle_name":       body.name.strip(),
        "number_plate":       body.vehicleNumber.strip(),
        "vehicle_type":       body.type,
        "capacity":           body.capacity,
        "assigned_driver_id": None,
        "created_at":         now,
        "updated_at":         now,
    }
    db = get_database()
    result = await db[settings.vehicles_collection].insert_one(doc)
    doc["_id"] = result.inserted_id
    logger.info(f"Vehicle created: {result.inserted_id}")
    return _serialise(doc)


@router.put("/{vehicle_id}")
async def update_vehicle(vehicle_id: str, body: VehicleUpdate):
    updates: dict = {"updated_at": datetime.now(timezone.utc)}
    if body.make is not None:
        updates["make"] = body.make.strip()
    if body.name is not None:
        updates["vehicle_name"] = body.name.strip()
    if body.vehicleNumber is not None:
        updates["number_plate"] = body.vehicleNumber.strip()
    if body.type is not None:
        if body.type not in VEHICLE_TYPES:
            raise HTTPException(status_code=422, detail=f"type must be one of {sorted(VEHICLE_TYPES)}")
        updates["vehicle_type"] = body.type
    if body.capacity is not None:
        updates["capacity"] = body.capacity
    if "assignedDriverId" in body.model_fields_set:
        updates["assigned_driver_id"] = body.assignedDriverId or None

    db = get_database()
    result = await db[settings.vehicles_collection].find_one_and_update(
        {"_id": _oid(vehicle_id)},
        {"$set": updates},
        return_document=True,
    )
    if result is None:
        raise HTTPException(status_code=404, detail="Vehicle not found")
    return _serialise(result)


@router.delete("/{vehicle_id}", status_code=204)
async def delete_vehicle(vehicle_id: str):
    db = get_database()
    result = await db[settings.vehicles_collection].delete_one({"_id": _oid(vehicle_id)})
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="Vehicle not found")

File: Backend-js/src/routes/vehicles.ts
import { Router } from 'express';
import { Vehicle } from '../models/index.js';
import { HttpError } from '../middleware/error.js';
import { requireRole } from '../middleware/auth.js';
const adminOnly = requireRole('super_admin', 'transportation_admin');
import { ah } from '../util/asyncHandler.js';
import { toOid } from '../util/oid.js';
import { logger } from '../logger.js';

const VEHICLE_TYPES = new Set(['SUV', 'Minivan', 'Van', 'Bus', 'Sedan', 'Truck']);
const OWNERSHIP_TYPES = new Set(['rented', 'volunteer_provided', 'sarthi_owned']);

function serialise(doc: Record<string, any>) {
  return {
    id: String(doc._id),
    make: doc.make ?? '',
    name: doc.vehicle_name ?? '',
    vehicleNumber: doc.number_plate ?? '',
    type: doc.vehicle_type ?? 'SUV',
    capacity: doc.capacity ?? 7,
    assignedDriverId: doc.assigned_driver_id ?? null,
    ownership: doc.ownership ?? 'rented',
    ownerName: doc.owner_name ?? '',
    ownerPhone: doc.owner_phone ?? '',
    ownerSarthiId: doc.owner_sarthi_id ?? null,
  };
}

export const vehiclesRouter = Router();

vehiclesRouter.get(
  '/',
  ah(async (_req, res) => {
    const docs = await Vehicle.find({}).lean();
    res.json(docs.map(serialise));
  })
);

vehiclesRouter.post(
  '/',
  adminOnly,
  ah(async (req, res) => {
    const body = req.body ?? {};
    const type: string = body.type ?? 'SUV';
    if (!VEHICLE_TYPES.has(type)) {
      throw new HttpError(422, `type must be one of ${JSON.stringify([...VEHICLE_TYPES].sort())}`);
    }
    const ownership: string = body.ownership ?? 'rented';
    if (!OWNERSHIP_TYPES.has(ownership)) {
      throw new HttpError(422, `ownership must be one of ${JSON.stringify([...OWNERSHIP_TYPES].sort())}`);
    }
    const now = new Date();
    const doc = {
      make: String(body.make ?? '').trim(),
      vehicle_name: String(body.name ?? '').trim(),
      number_plate: String(body.vehicleNumber ?? '').trim(),
      vehicle_type: type,
      capacity: body.capacity ?? 7,
      assigned_driver_id: null,
      ownership,
      owner_name: String(body.ownerName ?? '').trim(),
      owner_phone: String(body.ownerPhone ?? '').trim(),
      owner_sarthi_id: body.ownerSarthiId ?? null,
      created_at: now,
      updated_at: now,
    };
    const created = await Vehicle.create(doc);
    logger.info(`Vehicle created: ${created._id}`);
    res.status(201).json(serialise(created.toObject()));
  })
);

vehiclesRouter.put(
  '/:vehicleId',
  adminOnly,
  ah(async (req, res) => {
    const body = req.body ?? {};
    const updates: Record<string, any> = { updated_at: new Date() };
    if (body.make !== undefined && body.make !== null) updates.make = String(body.make).trim();
    if (body.name !== undefined && body.name !== null) updates.vehicle_name = String(body.name).trim();
    if (body.vehicleNumber !== undefined && body.vehicleNumber !== null)
      updates.number_plate = String(body.vehicleNumber).trim();
    if (body.type !== undefined && body.type !== null) {
      if (!VEHICLE_TYPES.has(body.type)) {
        throw new HttpError(422, `type must be one of ${JSON.stringify([...VEHICLE_TYPES].sort())}`);
      }
      updates.vehicle_type = body.type;
    }
    if (body.capacity !== undefined && body.capacity !== null) updates.capacity = body.capacity;
    if ('assignedDriverId' in body) updates.assigned_driver_id = body.assignedDriverId || null;
    if (body.ownership !== undefined && body.ownership !== null) {
      if (!OWNERSHIP_TYPES.has(body.ownership)) {
        throw new HttpError(422, `ownership must be one of ${JSON.stringify([...OWNERSHIP_TYPES].sort())}`);
      }
      updates.ownership = body.ownership;
    }
    if (body.ownerName !== undefined && body.ownerName !== null) updates.owner_name = String(body.ownerName).trim();
    if (body.ownerPhone !== undefined && body.ownerPhone !== null) updates.owner_phone = String(body.ownerPhone).trim();
    if ('ownerSarthiId' in body) updates.owner_sarthi_id = body.ownerSarthiId || null;

    const result = await Vehicle.findOneAndUpdate({ _id: toOid(req.params.vehicleId, 'vehicle id') }, { $set: updates }, { new: true }).lean();
    if (!result) throw new HttpError(404, 'Vehicle not found');
    res.json(serialise(result));
  })
);

vehiclesRouter.delete(
  '/:vehicleId',
  adminOnly,
  ah(async (req, res) => {
    const result = await Vehicle.deleteOne({ _id: toOid(req.params.vehicleId, 'vehicle id') });
    if (result.deletedCount === 0) throw new HttpError(404, 'Vehicle not found');
    res.status(204).end();
  })
);
